Thread-safe implementation of ReturnMultipleValues.

jbialobr opened the pull request:
Fixes #282.

zvirja commented:
@jbialobr Please rebase your code on top of the latest master - issues with CI were fixed there.

I've checked the suggested implementation and while it looks correct, this index arithmetic looks a bit messy to understand :)

Wouldn't it be much easier to use the `ConcurrentQueue<>` instead? So code will look like:
```csharp
public ReturnMultipleValues(T[] values)
{
	_lastValue = values.LastOrDefault();
	_valuesToReturn = new ConcurrentQueue<T>();
}

internal T GetNext()
{
	if (_valuesToReturn.TryDequeue(out T nextResult))
	{
		return nextResult;
	}
	else
	{
		return _lastValue;
	}
}
```

But of course, your code might work quicker and have less overhead for runtime - so it's a question :)

jbialobr commented:
> But of course, your code might work quicker and have less overhead for runtime - so it's a question :)

Yes, you are right, we should not see any difference in performance while performing a typical unit test.

Rebased and replaced with ConcurrentQueue.

zvirja commented on Source/NSubstitute/SubstituteExtensions.cs:
```diff
@@ -142,7 +142,7 @@ internal static ConfiguredCall Returns<T>(MatchArgs matchArgs, T returnThis, par
             else
             {
-                returnValue = new ReturnMultipleValues<T>(new[] { returnThis }.Concat(returnThese));
+                returnValue = new ReturnMultipleValues<T>(new[] { returnThis }.Concat(returnThese).ToArray());
```
Please pass an `IEnumerable` here and later make array in constructor if you need that. It's implementation detail that IEnumerables should be converted to arrays.

Moreover, we don't even need that - just initialize `ConcurrentQueue` with `IEnumerable` and later call `LastOrDefault()` on the created queue.

jbialobr replied:
I think that since ReturnMultipleValues is backed by ConcurrentQueue it should not accept IEnumerable as an argument. You can create an IEnumerable instance that produces infinite number of values.

zvirja replied:
Very interesting concern :)

I believe that should not be the case as I'd consider infinite enumerables as something really awful in design. If that would be a real concern, we would not be able to consume `IEnumerables` anywhere in the code as we often convert IEnumerables to arrays.
But let's see what @alexandrnikitin or @dtchepak think about this.

alexandrnikitin replied:
I prefer to limit the definition to more strict types, array in this case. `IEnumerable<T>` means not-materialized-in-memory/ undefined stream of data. The method taking `IEnumerable<T>` as an argument should take that into account and be lazy all the way down. We have finite and defined number of arguments in our case.

zvirja replied:
Thanks for your reply.
My concern was actually that we create redundant array which is abandoned immediately. I thought that it doesn't make sense to spam objects in places where we could easily omit that via design. 

Nevertheless, this extension method is invoked rarely, so we could disregard that :)

alexandrnikitin replied:
Well, regarding performance, there are other bottlenecks we can address ;) For instance #248

zvirja commented on Source/NSubstitute.Acceptance.Specs/MultipleThreads.cs:
```diff
@@ -112,6 +112,35 @@ public void Create_Delegate_Substitute_From_Many_Threads()
+        {
+            var substitute = Substitute.For<IFoo>();
+            System.Threading.Tasks.Task<int>[] tasks = new System.Threading.Tasks.Task<int>[10];
+            int[] substResults = new int[tasks.Length];
```
Could be simplified to `Enumerable.Range(..).ToArray()`

zvirja commented on Source/NSubstitute.Acceptance.Specs/MultipleThreads.cs:
```diff
@@ -112,6 +112,35 @@ public void Create_Delegate_Substitute_From_Many_Threads()
             Task.AwaitAll(tasks);
         }
 
+        [Test]
```
Please add one more test to test that the last value is returned if you ask more values than configured.

jbialobr replied:
Added an additional assertion for that.

jbialobr replied:
>It isn't a very good practice to have have multiple different assertions within the single test - it's harder to read them.

I would say that it is not a good practice to have multiple different assertions within a single test when the additional assertions make the test harder to read it.

zvirja replied:
ok, let's keep that up to @alexandrnikitin and @dtchepak.

alexandrnikitin replied:
@zvirja I think it's enough to have only multi-threaded test in our case. We have pletty of tests for `Returns(...)`

dtchepak commented:
Merged in, thanks a lot @jbialobr! Thanks @alexandrnikitin and @zvirja too for the PR review work.

dtchepak commented:
Released in 2.0.3.
